docs/basics/networkvariable.md

```diff
@@ -504,6 +504,51 @@ In order to create your own `NetworkVariableBase` derived container, you should:
+
+The first way is to use `GenerateSerializationForTypeAttribute` to generate serialization for a specific hard-coded type:
+```csharp
+[GenerateSerializationForType(typeof(Foo))]
+public class MyNetworkVariableTypeUsingFoo : NetworkVariableBase {}
+```
+
+However, this is rarely useful for generic types, since if you know the type, you can generally use the more efficient route of calling `FastBufferReader`/`FastBufferWriter` methods directly.
+
+The more common and more useful method is to use `GenerateSerializationForGenericParameterAttribute` to generate serialization for a specific generic parameter in your `NetworkVariable` type:
+```csharp
+[GenerateSerializationForGenericParameter(0)]
+public class MyNetworkVariableType<T> : NetworkVariableBase {}
+```
+
+This attribute takes an integer indicating which parameter in the type to generate serialization for. This value is 0-indexed, so the first type is 0, the second is 1, and so on. The attribute can be placed more than once on one class to generate serialization for multiple types:
+
+```csharp
+[GenerateSerializationForGenericParameter(0)]
+[GenerateSerializationForGenericParameter(1)]
+public class MyNetworkVariableType<TFirstType, TSecondType> : NetworkVariableBase {}
+```
+
+This will generate serialization for both `TFirstType` and `TSecondType`.
+
+Either of these attributes will cause the code generation to generate the following methods that you can use:
+
+```csharp
+NetworkVariableSerialization<T>.Write(FastBufferWriter writer, ref T value);
+NetworkVariableSerialization<T>.Read(FastBufferWriter writer, ref T value);
+NetworkVariableSerialization<T>.Duplicate(in T value, ref T duplicatedValue);
+NetworkVariableSerialization<T>.AreEqual(in T a, in T b);
+```
+
+In cases where the type is dynamically allocated (managed types and collections like NativeArray and NativeList), calling `Read` will read the value in-place whenever possible, avoiding further allocations.
+
+`AreEqual` can be used to determine if a value has changed from a previous value cached via `Duplicate` at a previous time to determine if a value has changed or not, to avoid re-sending the same value multiple times. This previous value cached via `Duplicate` can also be used to calculate deltas for implementing `ReadDelta` and `WriteDelta`.
+
+Note that the type must be serializable according to the "Supported Types" list above. Each type needs its own serializer instantiated, so this step tells the codegen which types to create serializers for.
+
+Note also that the codegen makes the assumption that all `NetworkVariable` types will exist as fields within `NetworkBehaviour` types, and limits its inspection to those fields to identify which types to create serializers for.
```

Comment:
What is "codegen" in this case? Is it just Unity's code generator / compiler?

Reply:
We have a step that runs after compiling using ILPP (IL Post-Processing) to inject additional automatically generated code into the assembly. That process scans the assembly for all types that inherit from NetworkBehaviour, inspects all the fields in those types to see if they are NetworkVariables, uses the attributes mentioned above to find out what kind of types those NetworkVariables will serialize, and then outputs code for each of the types it found to make them serializable based on various attributes of the type (is it a class or a struct, is it an INetworkSerializable type, is it an INetworkSerializeByMemcpy type, is it a C# built-in type, and so on)

Reply:
The user can interact with the process by using the `GenerateSerializationFor*` attributes, but otherwise, it is automatic and happens behind the scenes. I personally think that it is useful for a user to know that the process exists and what it does, but they do not need to know the details. Personally, when I am a user of something that feels like "magic", I want it to be documented what that magic is and does so I can understand how the system I'm using works, at least at a high level.

Reply:
Thanks for explaining that :) Appreciate it a lot. I mostly needed to know what to call that process so I've gone with "code generator" in my suggestion.